Fix csBestBlock/cvBlockChange waiting in rpc/mining

LarryRuane opened the pull request:
Cherry-pick https://github.com/bitcoin/bitcoin/pull/12743
This is an alternative to #5591, which isn't correct because it calls `chainActive->Tip()` without holding `cs_main`. The upstream fix is correct.

nuttycom commented on src/main.cpp:
```diff
@@ -3757,7 +3759,12 @@ void static UpdateTip(CBlockIndex *pindexNew, const CChainParams& chainParams) {
 
-    cvBlockChange.notify_all();
+    {
+        boost::unique_lock<boost::mutex> lock(g_best_block_mutex);
```
What is the difference between a `boost::unique_lock<boost::mutex>` and the `WaitableLock` type of the cherry-picked commit?

nuttycom commented on src/main.cpp:
```diff
@@ -66,8 +66,10 @@ BlockMap mapBlockIndex;
-CConditionVariable cvBlockChange;
+CWaitableCriticalSection g_best_block_mutex;
+CConditionVariable g_best_block_cv;
+uint256 g_best_block;
```
Since this adds a couple new global variables, please document them thoroughly, including where they are intended to be used. It'd be even nicer if we could figure out a way to avoid them being globals, but I don't see any obvious value that they could be tied to instead, so documentation will have to suffice.

LarryRuane replied:
I think cherry-picked commits should be as close to the source commits as possible (minimal diff) so that reviewers can see that they're at least functionally identical; then we can make further improvements (for example, in this case, not having these be global) in a separate PR. However, I do agree that adding documentation to a cherry-pick (beyond what's in the original commit) is fine; I'll do that here, thanks for the suggestion.

nuttycom replied:
I agree; can you add those as a separate commit?

nuttycom commented on src/main.h:
```diff
@@ -160,8 +160,13 @@ extern BlockMap mapBlockIndex;
-extern CWaitableCriticalSection csBestBlock;
-extern CConditionVariable cvBlockChange;
+
+// These prevent lock-ordering problems in getblocktemplate() RPC
```
Please also describe how these variables are intended to interact with one another.

nuttycom reviewed (dismissed):
utACK with request for better documentation about the newly added global variables. I have verified that this seems to reproduce the behavior of bitcoin/bitcoin#12743, though there are a number of differences in context.

daira commented on src/main.cpp:
```diff
@@ -3757,7 +3759,12 @@ void static UpdateTip(CBlockIndex *pindexNew, const CChainParams& chainParams) {
+    {
+        boost::unique_lock<boost::mutex> lock(g_best_block_mutex);
+        g_best_block = pindexNew->GetBlockHash();
+        g_best_block_height = pindexNew->nHeight;
```
I don't see an equivalent to this line in #12743. Where is the corresponding upstream change?

daira replied:
Oh I see, we need the height and upstream doesn't.

daira reviewed (commented):
~~Reviewed with question.~~ Answered my own question.

daira reviewed (approved):
utACK

LarryRuane commented:
Rebased (probably didn't need to), and addressed review comments as a new commit (so that the cherry-picks would be as close as possible to their source commits).

nuttycom reviewed (approved):
utACK

daira reviewed (commented):
Post-hoc ACK
